File: week1/Ariling/boj-1926-그림/main.js
function solution(input) {
  let check = [...Array(matrixInfo[0])].map(() =>
    Array(matrixInfo[1]).fill(false)
  );
  let count = 0;
  let max_width = 0;
  function bfs(i, j) {
    check[i][j] = true;
    let q = [[i, j]];
    let dr = [-1, 1, 0, 0];
    let dc = [0, 0, -1, 1];
    let num = 0;
    while (q.length > 0) {
      [r, c] = q.shift();
      num++;
      for (let x = 0; x < 4; x++) {
        let newR = r + dr[x];
        let newC = c + dc[x];
        if (
          newR >= 0 &&
          newR < matrixInfo[0] &&
          newC >= 0 &&
          newC < matrixInfo[1] &&
          !check[newR][newC] &&
          input[newR][newC] === 1
        ) {
          check[newR][newC] = true;
          q.push([newR, newC]);
        }
      }
    }
    return num;
  }
  for (let n = 0; n < matrixInfo[0]; n++) {
    for (let m = 0; m < matrixInfo[1]; m++) {
      if (!check[n][m] && arr[n][m] === 1) {
        width = bfs(n, m);
        if (width > max_width) {
          max_width = width;
        }
        count++;
      }
    }
  }
  console.log(count);
  console.log(max_width);
}

const readline = require("readline");

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let matrixInfo = [];
let arr = [];

rl.on("line", (line) => {
  if (matrixInfo.length === 0) {
    matrixInfo = line.split(" ").map(Number);
  } else {
    arr.push(line.split(" ").map(Number));
    if (arr.length === matrixInfo[0]) {
      rl.close();
    }
  }
}).on("close", () => {
  solution(arr);
  process.exit();
});

File: week1/Ariling/boj-1926-그림/main.py
import sys
input = sys.stdin.readline
from collections import deque
R, C = map(int, input().split())
arr = [list(map(int, input().split())) for _ in range(R)]
check_arr = [[False] * C for _ in range(R)]
# 그림 갯수, 그림 최대 넓이
result_arr = [0, 0]

def bfs(i, j):
    q = deque([])
    q.append((i, j))
    check_arr[i][j] = True
    dr = [-1, 1, 0, 0]
    dc = [0, 0, -1, 1]
    cnt = 0
    while q:
        r, c = q.popleft()
        cnt += 1
        for k in range(4):
            newR = r + dr[k]
            newC = c + dc[k]
            if 0 <= newR < R and 0 <= newC < C and not check_arr[newR][newC] and arr[newR][newC] == 1:
                check_arr[newR][newC] = True
                q.append((newR, newC))
    return cnt
for i in range(R):
    for j in range(C):
        if not check_arr[i][j] and arr[i][j] == 1:
            width = bfs(i, j)
            result_arr[0] += 1
            result_arr[1] = max(result_arr[1], width)
for i in range(len(result_arr)):
    print(result_arr[i])